### ISSUE TITLE
storage: use 1 RocksDB column family per table

### ISSUE THREAD

**Issue by ddorian** [cockroachdb/cockroach#7609]
Since you store "table_id/index_id" on each key, therefore on each block-index, and the block-index ~must be in-memory so you can find fast the right blocks, and the shorter the index the more data you can fit in 1 node, I think it makes sense to compress the keys  or maybe even keep them in a Map(table_id/index_id, Map(block_index)) ?

Makes sense ?

**Comment by petermattis** [comment]
@ddorian RocksDB, the low-level KV storage engine we use, performs prefix compressions which is a generalized form of the optimization you're describing. That's assuming I've understood you correctly.

**Comment by ddorian** [comment]
@petermattis Yes it does. Underneath data is separated in blocks of xKB and each block is compressed. 

For each block is saved the start+end key separately so you know which block to read from and decompress (calling this the block-index). 

So I mean prefix-compressing the block-index keys, since each range of X MB, will have only blocks of 1 table/index (because you split ranges when a new table begins).

Say 1 range of 64MB will have 1024 blocks of 64KB, and ALL the block-index-keys will have table-id/index-id the same, so maybe it makes sense to not repeat the table-id/index-id on the block-index in a range since it will always be the same ?

**Comment by ddorian** [comment]
Also, since a range will have only data about one table/index, then all the keys inside that range will have table-id/index-id prepended. Even though this is compressable, it will still consume space on ram before writing immutable block to disk and on commitlog. 

So it's probably better to remove the table-id/index-id also from the keys and not just from the block-index.
This way the table-id/index-id will only be defined once on each range.

**Comment by petermattis** [comment]
@ddorian I don't recall offhand if the block-index is prefix compressed or not, but assuming it is not, I'm not sure if it would be too important if it was. Blocks are currently configured to be 32KB in size while keys are on the O(100) bytes. Saving 2-3 bytes per key in the index would likely be in the noise.

Note that all of the keys in a CockroachDB cluster are in a "flat" namespace. And for a particular node, they are stored in a single RocksDB instance. Your proposals are reasonable at a high-level, but would require significant surgery either to RocksDB or our usage of RocksDB (i.e. using a RocksDB instance per range). Put another way, I don't see any way to remove the table-id/index-id from keys given our current usage of RocksDB. Please enlighten me if I'm missing something.

**Comment by ddorian** [comment]
@petermattis I thought you were using 1 db/range.

Rocksdb has [plugable memtable](https://github.com/facebook/rocksdb/wiki/RocksDB-Basics#memtables). The default one is a skip-list. You can create a new memtable implemented like ex: Map<table_id, skiplist>. Each table will have 1 skiplist and you'll repeat table-id only once on each memtable.

Rocksdb has [plugable db-format](https://github.com/facebook/rocksdb/wiki/Rocksdb-Table-Format). 
This too can be changed so in a ssfile you repeat the table-id only once. You can either separate blocks by "table-id", thus having a block-index for each table-id OR keep the table-id on the block-index and not in the raw blocks. This may(?) also make the bloom-filters smaller since they are for each block.

I don't think the commit-log is plugable (didnt find anything in the docs). If it is, that too can be changed to group keys so in 1 commit you write the table-id only once for each table.

This looks like "significant surgery". But maybe it's worth it...

**Comment by petermattis** [comment]
We use 1 RocksDB instance per store. A store is usually a disk and contains thousands of ranges. Plugging in a custom memtable implementation and db-format would be "significant surgery" in my opinion. My guess is that the benefit would a few percent space savings at most. That doesn't seem worth it to me.

**Comment by ddorian** [comment]
@petermattis 
Well, you can test the [HashSkipList](https://github.com/facebook/rocksdb/wiki/Hash-based-memtable-implementations) which is already done and see if it makes sense. Though it doesn't have the optimization to remove <table_id> from the keys on the memtable, just separating they keys by table-id. And it's 350 LOC in implementation (to modify and remove table-id if it makes sense https://github.com/facebook/rocksdb/blob/master/memtable/hash_skiplist_rep.cc )

Another way would be to use  [1-Column Family per (table-id](https://github.com/facebook/rocksdb/wiki/Column-Families). Column-families share the wal (slowest part on writes) and have separate memtables/sstables.

**Comment by petermattis** [comment]
@ddorian I'm aware of the `HashSkipList` memtable, but it isn't quite as simple as flipping a flag. `HashSkipList` hashes no the key prefix using the prefix-extractor, but we configure our prefix-extractor to pull out the "row" from a key. We could create a custom version of `HashSkipList` that extracted the table-id. I'm doubtful that will provide any meaningful memory savings or performance improvement (the memtable is currently configured to use 32 MB while the block cache is often gigabytes), but I'd be happy to guide someone else in performing that experiment.

Using a different column family per table-id is an interesting idea. There are complexities in doing this, primarily historical in nature. For example, when a table is created we don't immediately create a range associated with it but instead wait for the table to be split off from adjacent key spaces. We might end up fixing this wart for other reasons. I'm doubtful that there will be a significant space savings from this approach, but dropping a table by dropping a column family would be very attractive.

**Comment by petermattis** [comment]
Some other difficulties/questions with using a column family per table-id:

* The column family id (which is internally generated by RocksDB) gets encoded into the write batch representation. We're planning on passing the write batch representation between ranges. Perhaps RocksDB could be extended so that we could specify the column family id instead of having RocksDB generate it. Then we could say the column family id is the same as the table id and the write batch representation could be used as-is across multiple nodes.

* Column families have to be created/opened before use which would require a bunch of additional plumbing. 

* Can RocksDB handle thousands of column families efficiently? 

* We'd have to adjust stats collection to collect stats for every column family. There could be thousands or 10s of thousands in a large cluster. 

Switching to 1 column family per table-id is almost certainly not something we're going to tackle in the near to medium term, but it's an idea worth preserving. I'm going to retitle this issue and mark it "later".

**Comment by tschottdorf** [comment]
> Can RocksDB handle thousands of column families efficiently?

I vaguely (mis?)remember Mark Callaghan mentioning that upwards of a few 100 column families, he wouldn't recommend it out of the box. Could be making that up though.

**Comment by ddorian** [comment]
Maybe it's not such a good idea to have that many column families:

From mysql-rocksdb-facebook "Do not create too many column families: 20 is good enough":
http://www.slideshare.net/matsunobu/myrocks-deep-dive/55

Mongodb also went to 1 column-family ~same thing https://jira.mongodb.org/browse/SERVER-17225 because parse were using too many collections->column-families.

**Comment by petermattis** [comment]
Closing because it is very unlikely that we'll switch to 1 column family per table-id at this point.
